php/base-stemmer.php

```diff
@@ -0,0 +1,524 @@
+<?php
+/**
+ * Manually converted from JavaScript:
+ * https://github.com/snowballstem/snowball/blob/master/javascript/base-stemmer.js
+ * 
+ * Build a stemmer from Snowball, as follows:
+ * ~$ ./snowball algorithms/english.sbl -php -o php/english-stemmer
+ */
+abstract class SnowballStemmer {
+
+    protected string $current = '';
+    protected int $cursor = 0;
+    protected int $limit = 0;
+    protected int $limit_backward = 0;
+    protected int $bra = 0;
+    protected int $ket = 0;
+    
+    
+    abstract public function stem():bool;
+
+    
+    protected function setCurrent( string $value ):void {
+        $this->current = $value;
+        $this->cursor = 0;
+        $this->limit = $this->currentLength();
+        $this->limit_backward = 0;
+        $this->bra = $this->cursor;
+        $this->ket = $this->limit;
+    }
+
+
+    protected function getCurrent():string {
+        return $this->current;
+    }
+
+
+    /**
+     * @param int[] $s
+     */
+    protected function in_grouping( array $s, int $min, int $max ):bool {
+        if( $this->cursor >= $this->limit ){
+            return false;
+        }
+        $ch = $this->currentCharCodeAt($this->cursor);
+        if ($ch > $max || $ch < $min){
+            return false;
+        }
+        $ch -= $min;
+        if( ($s[ self::unsignedShiftRight($ch,3) ] & (0x1 << ($ch & 0x7))) === 0 ){
```

Comment:
The values being shifted are all in the range 0 to 255 so you can just use `>>` here.